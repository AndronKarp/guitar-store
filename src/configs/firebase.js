import firebase from "firebase";

const config = {
  apiKey: "AIzaSyDrF2CjklO0SlyV6Kv7pg7p67a4sLV3qbw",
  authDomain: "project-a0c8e.firebaseapp.com",
  databaseURL: "https://project-a0c8e.firebaseio.com",
  projectId: "project-a0c8e",
  storageBucket: "project-a0c8e.appspot.com",
  messagingSenderId: "686927330962",
  appId: "1:686927330962:web:6892625786538e6588304b",
  measurementId: "G-P6Q64LKV8Q"
};

firebase.initializeApp(config);

const database = firebase.database();
const guitarsRef = database.ref("guitars");
const auth = firebase.auth();
const usersRef = database.ref("users");
auth.getCurrentUser = () => {
  return new Promise((resolve, reject) => {
    const unsubscribe = firebase.auth().onAuthStateChanged(user => {
      unsubscribe();
      resolve(user);
    }, reject);
  });
};

export { guitarsRef, auth, usersRef };
