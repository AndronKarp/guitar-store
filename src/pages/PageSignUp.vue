<template>
  <div class="d-flex flex-fill justify-content-center align-items-center">
    <b-form
      @submit.prevent="signUp"
      class="d-flex flex-column w-25"
      style="min-width: 280px"
    >
      <b-form-group v-for="(field, index) in Object.keys(form)" :key="index">
        <b-form-input
          :id="form[field].id"
          :placeholder="form[field].placeholder"
          :type="form[field].type"
          v-model="$v.form[field].value.$model"
          :state="validationState(field)"
        ></b-form-input>
        <b-form-invalid-feedback :state="validationState(field)">
          {{ validationErrorMessage(field) }}
        </b-form-invalid-feedback>
      </b-form-group>

      <b-button
        :class="{ 'bg-info': !$v.$invalid }"
        :disabled="$v.$invalid"
        type="submit"
      >
        <b-spinner
          v-if="isFormSubmitting"
          label="Loading..."
          variant="light"
          small
        ></b-spinner>
        <template v-else>Sign Up</template>
      </b-button>
    </b-form>
  </div>
</template>

<script>
import { validationMixin } from "vuelidate";
import * as validators from "vuelidate/lib/validators";
import { auth, usersRef } from "../configs/firebase";
import { mapGetters } from "vuex";
import moveToHome from "../mixins/move-to-home";

export default {
  data() {
    return {
      form: {
        name: {
          id: "name",
          type: "text",
          placeholder: "Your name...",
          validations: {
            required: {
              rule: validators.required,
              errorMessage: "Required field"
            },
            alpha: {
              rule: validators.alpha,
              errorMessage: "Name must contain only letters"
            },
            minLength: {
              rule: validators.minLength(2),
              errorMessage: "Name must have at least 2 characters"
            }
          },
          value: ""
        },
        email: {
          id: "email",
          type: "email",
          placeholder: "Your e-mail...",
          validations: {
            required: {
              rule: validators.required,
              errorMessage: "Required field"
            },
            email: {
              rule: validators.email,
              errorMessage: "Invalid e-mail"
            },
            unique: {
              rule(email) {
                return !this.isEmailTaken(email);
              },
              errorMessage: "E-mail is already taken"
            }
          },
          value: ""
        },
        password: {
          id: "password",
          type: "password",
          placeholder: "Your password...",
          validations: {
            required: {
              rule: validators.required,
              errorMessage: "Required field"
            },
            alphaNum: {
              rule: validators.alphaNum,
              errorMessage: "Password must contain only letters and/or numbers"
            },
            minLength: {
              rule: validators.minLength(6),
              errorMessage: "Password must contain at least 6 characters"
            }
          },
          value: ""
        },
        confirmPassword: {
          id: "confirm-password",
          type: "password",
          placeholder: "Confirm password...",
          validations: {
            required: {
              rule: validators.required,
              errorMessage: "Required field"
            },
            sameAsPassword: {
              rule: validators.sameAs(function() {
                return this.form.password.value;
              }),
              errorMessage: "Passwords don't match"
            }
          },
          value: ""
        }
      },
      isFormSubmitting: false
    };
  },
  validations() {
    const form = {};
    const formFields = Object.entries(this.form);
    formFields.forEach(([fieldName, field]) => {
      const value = {};
      const validations = Object.entries(field.validations);
      validations.forEach(([validationName, validation]) => {
        value[validationName] = validation.rule;
      });
      form[fieldName] = { value };
    });
    return { form };
  },
  computed: {
    ...mapGetters(["currentUser", "isEmailTaken"]),
    validationState: () =>
      function(field) {
        const { $dirty, $error } = this.$v.form[field].value;
        return $dirty ? !$error : null;
      },
    validationErrorMessage: () =>
      function(field) {
        const errors = [];
        if (!this.$v.form[field].value.$dirty) return errors;
        const validations = Object.entries(this.form[field].validations);
        validations.forEach(([name, validation]) => {
          !this.$v.form[field].value[name] &&
            errors.push(validation.errorMessage);
        });
        return errors.length ? errors[errors.length - 1] : null;
      }
  },
  methods: {
    async signUp() {
      this.isFormSubmitting = true;
      await auth.createUserWithEmailAndPassword(
        this.form.email.value,
        this.form.password.value
      );
      await this.$store.dispatch("updateUserDisplayName", this.form.name.value);
      this.isFormSubmitting = false;
      this.moveToHomePage();
      this.addNewUserToDatabase();
    },
    addNewUserToDatabase() {
      usersRef.child(this.currentUser.uid).set({
        name: this.currentUser.displayName,
        email: this.currentUser.email
      });
    }
  },
  mixins: [validationMixin, moveToHome]
};
</script>

<style></style>
