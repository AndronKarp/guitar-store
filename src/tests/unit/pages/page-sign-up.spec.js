import { createLocalVue, mount } from "@vue/test-utils";
import PageSignUp from "@/pages/PageSignUp";
import { BootstrapVue } from "bootstrap-vue";
import Vuelidate from "vuelidate";
import Vuex from "vuex";
import { auth, unavailableEmailsRef } from "@/configs/firebase";

const localVue = createLocalVue();

localVue.use(BootstrapVue);
localVue.use(Vuelidate);
localVue.use(Vuex);

describe("PageSignUp.vue", () => {
  let wrapper;
  let actions;
  let store;
  let fetchUnavailableEmails;

  beforeEach(() => {
    unavailableEmailsRef.on = jest.fn((eventType, callback) =>
      callback({
        val: () => {
          return { value: "e-mail" };
        }
      })
    );
    actions = { updateUserDisplayName: jest.fn() };
    store = new Vuex.Store({
      actions
    });
    fetchUnavailableEmails = jest.spyOn(
      PageSignUp.methods,
      "fetchUnavailableEmails"
    );
    wrapper = mount(PageSignUp, {
      localVue,
      store
    });
  });

  test("renders registration form", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("computed", () => {
    const exampleField = "name";
    let exampleInput;

    describe("isEmailTaken", () => {
      test("returns true if unavailableEmails array contains passed email", () => {
        expect(wrapper.vm.isEmailTaken("e-mail")).toBe(true);
      });

      test("returns false if unavailableEmails array doesn't contain passed email", () => {
        expect(wrapper.vm.isEmailTaken("unique")).toBe(false);
      });
    });

    describe("validationState", () => {
      beforeEach(() => {
        exampleInput = wrapper.find(`#${exampleField}`);
      });

      test("returns null if input is clear", () => {
        expect(wrapper.vm.validationState(exampleField)).toBe(null);
      });

      test("returns false if input is dirty", () => {
        exampleInput.setValue("a");
        exampleInput.setValue("");
        expect(wrapper.vm.validationState(exampleField)).toBe(false);
      });

      test("returns false if input data is invalid", () => {
        exampleInput.setValue("a");
        expect(wrapper.vm.validationState(exampleField)).toBe(false);
      });

      test("returns true if input data is valid", () => {
        exampleInput.setValue("Andrew");
        expect(wrapper.vm.validationState(exampleField)).toBe(true);
      });
    });

    describe("validationErrorMessage", () => {
      beforeEach(() => {
        exampleInput = wrapper.find(`#${exampleField}`);
      });

      test("returns null if input data is valid", () => {
        exampleInput.setValue("Andrew");
        expect(wrapper.vm.validationErrorMessage(exampleField)).toBe(null);
      });

      test("returns error message of failed validation if input data is invalid", () => {
        exampleInput.setValue("a1");
        expect(wrapper.vm.validationErrorMessage(exampleField)).toBe(
          wrapper.vm.form[exampleField].validations.alpha.errorMessage
        );
      });
    });
  });

  describe("form submit button", () => {
    let submitButton;
    let inputs;

    beforeEach(() => {
      submitButton = wrapper.find("button");
      inputs = wrapper.findAll("input");
    });

    test("disabled if form is empty", () => {
      expect(submitButton.element.getAttribute("disabled")).toBe("disabled");
    });

    test("disabled if input data is invalid", () => {
      inputs.setValue("invalid");
      wrapper.vm.$nextTick(() => {
        expect(submitButton.element.getAttribute("disabled")).toBe("disabled");
      });
    });

    test("enabled if input data is valid", () => {
      inputs.wrappers[0].setValue("Andrew");
      inputs.wrappers[1].setValue("andrewkarp00@gmail.com");
      inputs.wrappers[2].setValue("123456");
      inputs.wrappers[3].setValue("123456");
      wrapper.vm.$nextTick(() => {
        expect(submitButton.element.getAttribute("disabled")).toBe(null);
      });
    });
  });

  describe("lifecycle methods", () => {
    test("created invokes fetchUnavailableEmails method", () => {
      expect(fetchUnavailableEmails).toHaveBeenCalled();
    });
  });

  describe("methods", () => {
    beforeEach(() => {
      unavailableEmailsRef.push = jest.fn();
    });

    test("fetchUnavailableEmails sets child added observer on unavailableEmails reference in database which pushes received data to unavailableEmails array", () => {
      expect(wrapper.vm.unavailableEmails).toContainEqual(
        expect.objectContaining({ value: "e-mail" })
      );
      expect(wrapper.vm.unavailableEmails).toHaveLength(1);
    });

    describe("signUp", () => {
      beforeEach(() => {
        auth.createUserWithEmailAndPassword = jest.fn();
        wrapper.vm.moveTo = jest.fn();
      });

      test("invokes auth.createUserWithEmailAndPassword function", async () => {
        await wrapper.vm.signUp();
        expect(auth.createUserWithEmailAndPassword).toHaveBeenCalledWith(
          wrapper.vm.form.email.value,
          wrapper.vm.form.password.value
        );
      });

      test("invokes updateUserDisplayName action", async () => {
        const dispatch = jest.spyOn(wrapper.vm.$store, "dispatch");
        await wrapper.vm.signUp();
        expect(dispatch).toHaveBeenCalledWith(
          "updateUserDisplayName",
          wrapper.vm.form.name.value
        );
      });

      test("invokes moveTo method", async () => {
        await wrapper.vm.signUp();
        expect(wrapper.vm.moveTo).toHaveBeenCalledWith("PageGuitarStore");
      });

      test("invokes saveNewUnavailableEmail method", async () => {
        const saveNewUnavailableEmail = jest.spyOn(
          wrapper.vm,
          "saveNewUnavailableEmail"
        );
        await wrapper.vm.signUp();
        expect(saveNewUnavailableEmail).toHaveBeenCalled();
      });
    });

    test("saveNewUnavailableEmail pushes passed object to database unavailableEmails reference", () => {
      wrapper.vm.saveNewUnavailableEmail();
      expect(unavailableEmailsRef.push).toHaveBeenCalledWith(
        expect.objectContaining({ value: wrapper.vm.form.email.value })
      );
    });
  });
});
