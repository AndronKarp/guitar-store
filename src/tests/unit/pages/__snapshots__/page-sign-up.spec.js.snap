// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSignUp.vue renders registration form 1`] = `
<div class="d-flex flex-fill justify-content-center align-items-center">
  <form novalidate="novalidate" class="d-flex flex-column w-25" style="min-width: 280px;">
    <fieldset class="form-group" id="__BVID__29">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="name" type="text" placeholder="Your name..." class="form-control">
        <div class="invalid-feedback">
          Required field
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__31">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="email" type="email" placeholder="Your e-mail..." class="form-control">
        <div class="invalid-feedback">
          Required field
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__33">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="password" type="password" placeholder="Your password..." class="form-control">
        <div class="invalid-feedback">
          Required field
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
    <fieldset class="form-group" id="__BVID__35">
      <!---->
      <div tabindex="-1" role="group" class="bv-no-focus-ring"><input id="confirm-password" type="password" placeholder="Confirm password..." class="form-control">
        <div class="invalid-feedback">
          Required field
        </div>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset> <button type="submit" disabled="disabled" class="btn btn-secondary disabled">Sign Up</button>
  </form>
</div>
`;
